### ISSUE TITLE
HLSL: Map min16float/min16int/min16uint to float16_t/int16_t/uint16_t

### ISSUE THREAD

**Issue by amdrexu** [KhronosGroup/glslang#1271]
Maybe add an option to conditionally map those types in HLSL to AMD float16_t/int16_t/uint16_t.  Reference the requirement https://github.com/KhronosGroup/glslang/pull/1266.

**Comment by johnkslang** [comment]
Perhaps options like these should be part of the --target-env options.

**Comment by amdrexu** [comment]
I make a pull request to address this requirement. I add an option "--hlsl-enable-16bit-types". I think it is inappropriate to add this to --target-env. The option is better to be an independent one. I want to enlarge the enum TOptions but it is full. Therefore, I have to add a boolean variable to represent this option.

**Comment by amdrexu** [comment]
See the pull request: https://github.com/KhronosGroup/glslang/pull/1275

**Comment by oscarbg** [comment]
many thanks for your (fast) support!
is really appreciated..

**Comment by oscarbg** [comment]
Hi @amdrexu ,
testing it still seems has some issues :
I'm attaching test.hlsl and all generated outputs here:
DOWNLOAD->
[tests.zip](https://github.com/KhronosGroup/glslang/files/1800468/tests.zip)

1)it works correctly on DXC  -enable-16bit-types 

> dxc.exe -E CSMandelJulia_scalarFloat test.hlsl -enable-16bit-types -T cs_6_2 -spirv -O0 -Vd >test.txt

2) of course works with glslangValidator default option of translating to float

> glslangValidator -D -H -e CSMandelJulia_scalarFloat -S comp test.hlsl >testglslc.txt

3)with -enable-16bit-types fails:

> glslangValidator -D -H -e CSMandelJulia_scalarFloat -S comp --hlsl-enable-16bit-types test.hlsl 

>testglslcfp16.txt
returns:
```
test.hlsl
ERROR: test.hlsl:32: '' : unsupported construction 
ERROR: test.hlsl:32: ' temp mediump float16_t' : cannot construct with these arguments 
ERROR: test.hlsl:32: 'assignment expression' : Expected 
test.hlsl(32): error at column 34, HLSL parsing failed.
ERROR: 4 compilation errors.  No code generated.
SPIR-V is not generated for failed compile or link
```

**Comment by amdrexu** [comment]
Reproduced. I will have a look and give you update.

**Comment by oscarbg** [comment]
Ok thanks..
tested already int16 shader:
glslangValidator -D -H -e CSMandelJulia_scalarFloat -S comp --hlsl-enable-16bit-types testint16.hlsl
I attach the two HLSL files **the original testint16ori.hlsl ** and the one doing similar min16int(2) manual casts : as testint16.hlsl
[testhlslint16.zip](https://github.com/KhronosGroup/glslang/files/1802240/testhlslint16.zip)
the fixed casts one works once adding similar fix (see below)
one comment:
then I see the SPIR-V:
         OpCapability Int16
           OpExtension "SPV_AMD_gpu_shader_int16"
similar to min16float case, is this OK?
I say because OpCapability Int16 is probably enough for all min16int ops HLSL supports or there are some ops that might need OpExtension "SPV_AMD_gpu_shader_int16" support?
perhaps should probably ask SPIRV-Tools devs also..

```

	case EOpConstructI8Vec2:
	case EOpConstructI8Vec3:
	case EOpConstructI8Vec4:
	case EOpConstructInt8:
		basicOp = EOpConstructInt8;
		break;

	case EOpConstructU8Vec2:
	case EOpConstructU8Vec3:
	case EOpConstructU8Vec4:
	case EOpConstructUint8:
		basicOp = EOpConstructUint8;
		break;

	case EOpConstructI16Vec2:
	case EOpConstructI16Vec3:
	case EOpConstructI16Vec4:
	case EOpConstructInt16:
		basicOp = EOpConstructInt16;
		break;

	case EOpConstructU16Vec2:
	case EOpConstructU16Vec3:
	case EOpConstructU16Vec4:
	case EOpConstructUint16:
		basicOp = EOpConstructUint16;
		break;

```

**Comment by amdrexu** [comment]
I have made the similar fix. Add missing constructor ops. If you specify SPIR-V 1.3 environment, then AMD extension string will not be generated. I also find a problem in your min16float shader, see this statment `b = b * 0.5f`. I think `0.5f` should be `min16float(0.5f)` or `0.5hf`. Otherwise, when min16float is mapped to float16, this statement is float16 = float16 * float. It is invalid because the destination b is float16 and 0.5f is float. Implicit conversion is disallowed in such case.

**Comment by amdrexu** [comment]
I make a fix here: https://github.com/KhronosGroup/glslang/pull/1288

**Comment by oscarbg** [comment]
thanks..
strange then than DirectXshaderCompiler SPIR-V compiled OK this shader.. seems more relaxed..
also what about the remaining case v= 2* (uv + b)?
i.e. for min16float=integer*min16float  is implicit conversion allowed in this case?
in case yes, your commit fixes this case?

**Comment by amdrexu** [comment]
Yes, I test your shader. testint16.hlsl works as expected with the change `min16int(2)` and testintori.hlsl fails for the same reason. I think this is because HLSL semantics parser in glslang is based on GLSL parser. In GLSL, implicit conversion is allowed when the precision or value range is improved and the table 4.1.10 Implicit Conversions in GLSL spec tells us which types could be promoted to which types. For int16 and float16, the conversion int -> int16 and float -> float16 both have precision loss. So such implicit conversions are disabled. As you said, DX compiler might be more relaxed.

**Comment by TheJackiMonster** [comment]
> Yes, I test your shader. testint16.hlsl works as expected with the change `min16int(2)` and testintori.hlsl fails for the same reason. I think this is because HLSL semantics parser in glslang is based on GLSL parser. In GLSL, implicit conversion is allowed when the precision or value range is improved and the table 4.1.10 Implicit Conversions in GLSL spec tells us which types could be promoted to which types. For int16 and float16, the conversion int -> int16 and float -> float16 both have precision loss. So such implicit conversions are disabled. As you said, DX compiler might be more relaxed.

Is there an option to allow such implicit conversions with Glslang's HLSL parser? Because I've encountered quite a few written shaders which use that a lot and it would be neat if I wouldn't need to adjust them (adding explicit conversion manually) every time. ^^'
